src/adapters/EasyRtcAdapter.js

```diff
@@ -177,19 +173,73 @@ class EasyRtcAdapter extends NoOpAdapter {
     }
   }
 
-  getMediaStream(clientId) {
-    var that = this;
-    if (this.audioStreams[clientId]) {
-      NAF.log.write("Already had audio for " + clientId);
-      return Promise.resolve(this.audioStreams[clientId]);
+  getMediaStream(clientId, type = "audio") {
+    if (this.mediaStreams[clientId]) {
+      NAF.log.write(`Already had ${type} for ${clientId}`);
+      return Promise.resolve(this.mediaStreams[clientId][type]);
     } else {
-      NAF.log.write("Waiting on audio for " + clientId);
-      return new Promise(function(resolve) {
-        that.pendingAudioRequest[clientId] = resolve;
-      });
+      NAF.log.write(`Waiting on ${type} for ${clientId}`);
+      if (!this.pendingMediaRequests.has(clientId)) {
+        this.pendingMediaRequests.set(clientId, {});
+
+        const audioPromise = new Promise((resolve, reject) => {
+          this.pendingMediaRequests.get(clientId).audio = { resolve, reject };
+        });
+        const videoPromise = new Promise((resolve, reject) => {
+          this.pendingMediaRequests.get(clientId).video = { resolve, reject };
+        });
+
+        this.pendingMediaRequests.get(clientId).audio.promise = audioPromise;
+        this.pendingMediaRequests.get(clientId).video.promise = videoPromise;
+
+        audioPromise.catch(e => NAF.log.warn(`${clientId} getMediaStream Audio Error`, e));
+        videoPromise.catch(e => NAF.log.warn(`${clientId} getMediaStream Video Error`, e));
+      }
+      return this.pendingMediaRequests.get(clientId)[type].promise;
     }
   }
 
+  setMediaStream(clientId, stream) {
+    // Safari doesn't like it when you use single a mixed media stream where one of the tracks is inactive, so we
+    // split the tracks into two streams.
+    const audioStream = new MediaStream();
+    try {
+      stream.getAudioTracks().forEach(track => audioStream.addTrack(track));
+    } catch(e) {
+      NAF.log.warn(`${clientId} setMediaStream Audio Error`, e);
+    }
+
+    const videoStream = new MediaStream();
+    try {
+      stream.getVideoTracks().forEach(track => videoStream.addTrack(track));
+    } catch (e) {
+      NAF.log.warn(`${clientId} setMediaStream Video Error`, e);
+    }
+
+    this.mediaStreams[clientId] = { audio: audioStream, video: videoStream };
+
+    // Resolve the promise for the user's media stream if it exists.
+    if (this.pendingMediaRequests.has(clientId)) {
+      this.pendingMediaRequests.get(clientId).audio.resolve(audioStream);
+      this.pendingMediaRequests.get(clientId).video.resolve(videoStream);
+    }
+  }
+
+  setLocalMediaStream(stream) {
+    this.setMediaStream(
+        this.easyrtc.myEasyrtcid,
+        stream
+    );
+  }
+
+  enableMicrophone(enabled) {
+    this.easyrtc.enableMicrophone(enabled);
+  }
+
+  enableCamera(enabled) {
+    this.easyrtc.enableCamera(enabled);
```

Comment:
Maybe calling ` this.easyrtc.enableVideo(true)` make sense here before calling enableCamera? This is to renegotiate the sdp offer right?